prov/shm/src/smr.h

```diff
@@ -324,6 +325,56 @@ uint64_t smr_rx_cq_flags(uint32_t op, uint16_t op_flags);
 
 void smr_ep_progress(struct util_ep *util_ep);
 
+static inline int smr_cma_loop_read(pid_t pid, struct iovec *local,
+			unsigned long local_cnt, struct iovec *remote,
+			unsigned long remote_cnt, unsigned long flags,
+			size_t total)
+{
+	ssize_t ret;
+
+	while (1) {
+		ret = ofi_process_vm_readv(pid, local, local_cnt, remote,
+					   remote_cnt, flags);
+		if (ret < 0) {
+			FI_WARN(&smr_prov, FI_LOG_EP_CTRL,
+				"CMA read error %d\n", errno);
+			return -FI_EIO;
+		}
+
+		total -= ret;
+		if (!total)
+			return FI_SUCCESS;
+
+		ofi_consume_iov(local, &local_cnt, (size_t) ret);
+		ofi_consume_iov(remote, &remote_cnt, (size_t) ret);
+	}
+}
+
+static inline int smr_cma_loop_write(pid_t pid, struct iovec *local,
+			unsigned long local_cnt, struct iovec *remote,
+			unsigned long remote_cnt, unsigned long flags,
+			size_t total)
+{
+	ssize_t ret;
+
+	while (1) {
+		ret = ofi_process_vm_writev(pid, local, local_cnt, remote,
+					    remote_cnt, flags);
+		if (ret < 0) {
+			FI_WARN(&smr_prov, FI_LOG_EP_CTRL,
+				"CMA write error %d\n", errno);
+			return -FI_EIO;
+		}
+
+		total -= ret;
+		if (!total)
+			return FI_SUCCESS;
+
+		ofi_consume_iov(local, &local_cnt, (size_t) ret);
+		ofi_consume_iov(remote, &remote_cnt, (size_t) ret);
+	}
+}
```

Comment:
The above 2 functions are nearly the same, except for calling read/write.  And everywhere they are called, there is a check for read/write.  Combine the two functions together and move the check inside.

Reply:
Ha, that's originally how I had it and I said to myself, "Sean would tell me to separate these so I'm not checking it every time"

Reply:
I thought about that, but since each iteration handles 2GB, I'm not sure how often the loop will ever be taken more than once with a real app.  Plus the overhead of a check will be swamped by the 2GB data copy anyway...

If you want to go all out, you could perform the check above the loop, and set a function pointer to read/write, which the loop will call.

Reply:
Nah, you're right. This doesn't happen enough to warrant the extra code for such a small optimization. I'll move it inside.

Reply:
Actually I realized the other reason I separated it - the checks are backwards in the two ways CMA is used (one is initiated by the sender on the fast path and one by the receiver so the same op initiates opposite CMA calls). Still want me to separate it?

Reply:
I would pass in a bool indicating read/write direction

Reply:
Hokay. Updated.